### ISSUE TITLE
EpiDISH

### ISSUE THREAD

**Issue by sjczheng** [Bioconductor/Contributions#401]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/sjczheng/EpiDISH

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by dvantwisk** [comment]
We've reviewed your package!  It looks good, but there are places for improvement.  

DESCRIPTION

- L8: Unless the user needs to have access to the symbols in the packages in
'Depends: ', it is best to place these in a section labeled 'Imports: '


NAMESPACE

- Consider explicitly enumerating what symbols are being imported using
`importFrom()` in the `roxygen2` header.


R

R/DummyBeta.m.R & R/centDHSblood.m.R
- Needs more description for man pages so a new user can understand what the
data sets are.

R/epidish.R

- L94: Better to put argument validation upfront instead of at end of function.
Use `match.arg()`. 

- L101: Similar to the issue with L94, use `match.arg()` to validate
constraints input.  Make 'constraints' in function definition equal to
c('inequality', 'equality').

- L124: Consider using `seq_len()` instead of `:` in the for-loop.  This prevents
the edge case where `1:0` results in `0`.

- L168-170: Code can be vectorized with:

```{r}
s <- seq_len(ncol(avdata.m))
rmse.m[s,nui] <- sqrt(colMeans((data2.m[,s] - reconst.m[,s])^2))
```

- L171: Use `message()` instead of print.  `message()` will output to stderr as
opposed to `print()` that outputs to stdout.

- L187-192: Vectorize the code.  Instead of nested for-loops, replace with:
```{r}
D <- apply(ref.m, 2, function(x) colSums(x * ref.m))
```

- L196-201: Simplify code with:
```{r}
A.m <- matrix(1, nrow=ncol(ref.m) + 1, ncol=nW)
A.m[,1] <- -1
```

- L194-247: The if and else blocks both have almost the same code.   Consider
making a private function that performs the task so that code is not repeated.

- Consider using the `formatR` package for cleaning up indentation in this file.


man

Looks good!


test

[Recommended] Include tests to verify functionality and input handling of the
`epidish()` function.  See the `testthat` package for more information.


vignettes

Looks good!

**Comment by sjczheng** [comment]
Hi, @dvantwisk 

Thank you so much for your comments. It helps a lot.
I

**Comment by sjczheng** [comment]
Hi, @dvantwisk 

Thank you so much for your comments. I have learned a lot from it. 
I've made the following improvements:

DESCRIPTION
L8: Using 'Imports:' instead of 'Depends'

NAMESPACE
Now using importFrom()

R

R/DummyBeta.m.R & R/centDHSblood.m.R
 More description added

R/epidish.R

L94: Putting argument validation upfront instead of at end of function. Using match.arg().

L101: Putting argument validation upfront instead of at end of function. Using match.arg().
. Making 'constraints' in function definition equal to c('inequality', 'equality').

All: Using seq_len() instead of : in the for-loop. 

L168-170: Vectorized with:
s <- seq_len(ncol(avdata.m))
rmse.m[s,nui] <- sqrt(colMeans((data2.m[,s] - reconst.m[,s])^2))

L171: Using message() instead of print. 

L187-192: Replacig with:
D <- apply(ref.m, 2, function(x) colSums(x * ref.m))

L196-201: **I cannot simplify the code as in your comments for this one. I have 0, 1, -1 in the matrix. Your comment here seems to give different result.**

L194-247: If and else chunks are more concise now.

I have used the formatR package for cleaning up indentation in this file.

test

Using testthat package to test the n. of columns consistency
